src/server.rs

```diff
@@ -259,26 +260,43 @@ impl<L, M> Server<L, M> where L: Log, M: StateMachine {
+                            let peer = peer.unwrap(); // TODO(tschottdorf): nicer way?
+                            let peer_id = ServerId(peer.get_id());
 
-                            self.connections[token].set_kind(ConnectionKind::Peer(peer_id));
-                            let prev_token = self.peer_tokens
-                                                 .insert(peer_id, token)
-                                                 .expect("peer token not found");
-
-                            // Close the existing connection.
-                            self.connections
-                                .remove(prev_token)
-                                .expect("peer connection not found");
+                            // Not the source address of this connection, but the
+                            // address the peer tells us it's listening on.
+                            let peer_addr = SocketAddr::from_str(try!(peer.get_addr())).unwrap();
+                            scoped_debug!("received new connection from {:?} ({})", peer_id, peer_addr);
 
-                            // Clear any timeouts associated with the existing connection.
-                            self.reconnection_timeouts
-                                .remove(&prev_token)
-                                .map(|handle| scoped_assert!(event_loop.clear_timeout(handle)));
+                            self.connections[token].set_kind(ConnectionKind::Peer(peer_id));
+                            // Use the advertised address, not the remote's source
+                            // address, for future retries in this connection.
+                            self.connections[token].set_addr(peer_addr);
 
-                            // TODO: add reconnect messages from consensus
+                            let prev_token = Some(self.peer_tokens
+                                                 .insert(peer_id, token)
+                                                 .expect("peer token not found"));
+
+                            // Close the existing connection, if any.
+                            // Currently, prev_token is never `None`; see above.
+                            // With config changes, this will have to be handled.
+                            match prev_token {
+                                Some(tok) => {
+                                    self.connections
+                                        .remove(tok)
+                                        .expect("peer connection not found");
+
+                                    // Clear any timeouts associated with the existing connection.
+                                    self.reconnection_timeouts
+                                        .remove(&tok)
+                                        .map(|handle| scoped_assert!(event_loop.clear_timeout(handle)));
+                                }
+                                _ => ()
```

Comment:
Can you make this `unreachable!()` instead of `()` so that we will know if/when it starts happening?  Alternatively you can just keep it as an implicit `unwrap`.